stratum/tools/stratum-replay/stratum_replay.cc

```diff
@@ -0,0 +1,229 @@
+  uint64 election_id_high;
+  uint64 election_id_low;
+  CHECK_RETURN_IF_FALSE(absl::SimpleAtoi(election_ids[0], &election_id_high))
+      << "Unable to parse string " << election_ids[0] << " to uint64";
+  CHECK_RETURN_IF_FALSE(absl::SimpleAtoi(election_ids[1], &election_id_low))
+      << "Unable to parse string " << election_ids[1] << " to uint64";
+  absl::uint128 election_id =
+      absl::MakeUint128(election_id_high, election_id_low);
+
+  stream_req.mutable_arbitration()->set_device_id(FLAGS_device_id);
+  stream_req.mutable_arbitration()->mutable_election_id()->set_high(
+      absl::Uint128High64(election_id));
+  stream_req.mutable_arbitration()->mutable_election_id()->set_low(
+      absl::Uint128Low64(election_id));
+
+  ::grpc::ClientContext context;
+  std::unique_ptr<ClientStreamChannelReaderWriter> stream =
+      stub->StreamChannel(&context);
+  if (!stream->Write(stream_req)) {
+    RETURN_ERROR(ERR_INTERNAL)
+        << "Failed to send request '" << stream_req.ShortDebugString()
+        << "' to switch.";
+  }
+
+  // Push the given pipeline config
+  ::p4::v1::SetForwardingPipelineConfigRequest fwd_pipe_cfg_req;
+  ::p4::v1::SetForwardingPipelineConfigResponse fwd_pipe_cfg_resp;
+  fwd_pipe_cfg_req.set_device_id(FLAGS_device_id);
+  fwd_pipe_cfg_req.mutable_election_id()->set_high(
+      absl::Uint128High64(election_id));
+  fwd_pipe_cfg_req.mutable_election_id()->set_low(
+      absl::Uint128Low64(election_id));
+  fwd_pipe_cfg_req.set_action(
+      ::p4::v1::SetForwardingPipelineConfigRequest::VERIFY_AND_COMMIT);
+
+  ::stratum::hal::ForwardingPipelineConfigs pipeline_cfg;
+  RETURN_IF_ERROR(ReadProtoFromTextFile(FLAGS_pipeline_cfg, &pipeline_cfg));
+  const ::p4::v1::ForwardingPipelineConfig* fwd_pipe_cfg =
+      ::stratum::gtl::FindOrNull(pipeline_cfg.node_id_to_config(),
+                                 FLAGS_device_id);
+  CHECK_RETURN_IF_FALSE(fwd_pipe_cfg);
+  fwd_pipe_cfg_req.mutable_config()->CopyFrom(*fwd_pipe_cfg);
+
+  ::grpc::Status status;
+  {
+    ::grpc::ClientContext context;
+    status = stub->SetForwardingPipelineConfig(&context, fwd_pipe_cfg_req,
+                                               &fwd_pipe_cfg_resp);
+    CHECK_RETURN_IF_FALSE(status.ok())
+        << "Failed to push forwarding pipeline config: "
+        << ::stratum::hal::P4RuntimeGrpcStatusToString(status);
+  }
+
+  // Parse the P4Runtime write log file and send write requests to the
+  // target device.
+  std::string p4WriteLogs;
+  RETURN_IF_ERROR(::stratum::ReadFileToString(argv[1], &p4WriteLogs));
+  std::vector<std::string> lines =
+      absl::StrSplit(p4WriteLogs, '\n', absl::SkipEmpty());
+  for (const std::string& line : lines) {
+    ::p4::v1::WriteRequest write_req;
+    ::p4::v1::WriteResponse write_resp;
+    // Log format: <timestamp>;<node_id>;<update proto>;<status>
+    // This regular expression contains 4 sub-match groups which extracts
+    // elements from the log string.
```

Comment:
I think we should reference `LogWriteRequest` in `P4Service` here, where the format is defined.